File: app/app/services/organization.js
// Copyright 2016 Documize Inc. <legal@documize.com>. All rights reserved.
//
// This software (Documize Community Edition) is licensed under
// GNU AGPL v3 http://www.gnu.org/licenses/agpl-3.0.en.html
//
// You can operate outside the AGPL restrictions by purchasing
// Documize Enterprise Edition and obtaining a commercial license
// by contacting <sales@documize.com>.
//
// https://documize.com

import Ember from 'ember';

const {
	inject: { service }
} = Ember;

export default Ember.Service.extend({
	sessionService: service('session'),
	ajax: service(),
	appMeta: service(),
	store: service(),

	// Returns attributes for specified org id.
	getOrg(id) {
		return this.get('ajax').request(`organizations/${id}`, {
			method: 'GET'
		}).then((response) => {
			let data = this.get('store').normalize('organization', response);
			return this.get('store').push({ data: data });
		});
	},

	// Updates an existing organization record.
	save(org) {
		let id = org.get('id');

		this.get('appMeta').setProperties({
			message: org.message,
			title: org.title
		});

		return this.get('ajax').request(`organizations/${id}`, {
			method: 'PUT',
			data: JSON.stringify(org)
		});
	}
});

File: domain/organization/organization.go
// Copyright 2016 Documize Inc. <legal@documize.com>. All rights reserved.
//
// This software (Documize Community Edition) is licensed under
// GNU AGPL v3 http://www.gnu.org/licenses/agpl-3.0.en.html
//
// You can operate outside the AGPL restrictions by purchasing
// Documize Enterprise Edition and obtaining a commercial license
// by contacting <sales@documize.com>.
//
// https://documize.com

package organization

import (
	"net/http"
	"strings"
)

// GetRequestSubdomain extracts subdomain from referring URL.
func GetRequestSubdomain(r *http.Request) string {
	return urlSubdomain(r.Referer())
}

// GetSubdomainFromHost extracts the subdomain from the requesting URL.
func GetSubdomainFromHost(r *http.Request) string {
	return urlSubdomain(r.Host)
}

func urlSubdomain(url string) string {
	url = strings.ToLower(url)
	url = strings.Replace(url, "https://", "", 1)
	url = strings.Replace(url, "http://", "", 1)

	parts := strings.Split(url, ".")

	if len(parts) >= 2 {
		url = parts[0]
	} else {
		url = ""
	}

	return url
}
